SignalServiceKit/src/Storage/FullTextSearchFinder.swift

```diff
@@ -0,0 +1,141 @@
+
+    public func index(_ item: T) -> String {
+        return indexBlock(item)
+    }
+}
+
+@objc
+public class FullTextSearchFinder: NSObject {
+
+    public func enumerateObjects(searchText: String, transaction: YapDatabaseReadTransaction, block: @escaping (Any, String) -> Void) {
+        guard let ext: YapDatabaseFullTextSearchTransaction = ext(transaction: transaction) else {
+            assertionFailure("ext was unexpectedly nil")
+            return
+        }
+
+        let normalized = FullTextSearchFinder.normalize(text: searchText)
+
+        // We want a forgiving query for phone numbers
+        // TODO a stricter "whole word" query for body text?
+        let prefixQuery = "*\(normalized)*"
+
+        // (snippet: String, collection: String, key: String, object: Any, stop: UnsafeMutablePointer<ObjCBool>)
+        ext.enumerateKeysAndObjects(matching: prefixQuery, with: nil) { (snippet: String, _: String, _: String, object: Any, _: UnsafeMutablePointer<ObjCBool>) in
+            block(object, snippet)
+        }
+    }
+
+    private func ext(transaction: YapDatabaseReadTransaction) -> YapDatabaseFullTextSearchTransaction? {
+        return transaction.ext(FullTextSearchFinder.dbExtensionName) as? YapDatabaseFullTextSearchTransaction
+    }
+
+    // Mark: Index Building
+
+    private class var contactsManager: ContactsManagerProtocol {
+        return TextSecureKitEnv.shared().contactsManager
+    }
+
+    private class func normalize(text: String) -> String {
+        var normalized: String = text.trimmingCharacters(in: .whitespacesAndNewlines)
+
+        // Remove any phone number formatting from the search terms
+        let nonformattingScalars = normalized.unicodeScalars.lazy.filter {
+            !CharacterSet.punctuationCharacters.contains($0)
+        }
+
+        normalized = String(String.UnicodeScalarView(nonformattingScalars))
+
+        return normalized
+    }
+
+    private static let groupThreadIndexer: SearchIndexer<TSGroupThread> = SearchIndexer { (groupThread: TSGroupThread) in
+        let groupName = groupThread.groupModel.groupName ?? ""
+
+        let memberStrings = groupThread.groupModel.groupMemberIds.map { recipientId in
+            recipientIndexer.index(recipientId)
+        }.joined(separator: " ")
+
+        let searchableContent = "\(groupName) \(memberStrings)"
+
+        return normalize(text: searchableContent)
+    }
+
+    private static let contactThreadIndexer: SearchIndexer<TSContactThread> = SearchIndexer { (contactThread: TSContactThread) in
+        let recipientId =  contactThread.contactIdentifier()
+        let searchableContent = recipientIndexer.index(recipientId)
+
+        return normalize(text: searchableContent)
+    }
+
+    private static let recipientIndexer: SearchIndexer<String> = SearchIndexer { (recipientId: String) in
+        let displayName = contactsManager.displayName(forPhoneIdentifier: recipientId)
+        let searchableContent =  "\(recipientId) \(displayName)"
```

Comment:
I wonder if we should include system contact names, profile names, phone number.

We'd want `+13213214321` and `(321) 321-4321)` to match `+13213214321`, right?

Reply:
> We'd want +13213214321 and (321) 321-4321) to match +13213214321, right?

Yes - there's a card for this.